### PR TITLE
make tests run on numpy1.x and numpy2.x

### LINKED ISSUE #6072: [BUG] Test failures with NumPy 2.0b1
### Describe the bug

Several tests fail with NumPy 2.0b1, many apparently due to a different `repr()` of NumPy types:
```
======================================================================
FAIL: one_dim (memoryview_annotation_typing)
Doctest: memoryview_annotation_typing.one_dim
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/tests/memoryview/memoryview_annotation_typing.py", line 19, in memoryview_annotation_typing.one_dim
Failed example:
    one_dim(a)
Expected:
    (2.0, 1)
Got:
    (np.float64(2.0), 1)


======================================================================
FAIL: one_dim_ccontig (memoryview_annotation_typing)
Doctest: memoryview_annotation_typing.one_dim_ccontig
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/tests/memoryview/memoryview_annotation_typing.py", line 29, in memoryview_annotation_typing.one_dim_ccontig
Failed example:
    one_dim_ccontig(a)
Expected:
    (2.0, 1)
Got:
    (np.float64(2.0), 1)


======================================================================
FAIL: one_dim_nogil_cfunc (memoryview_annotation_typing)
Doctest: memoryview_annotation_typing.one_dim_nogil_cfunc
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/tests/memoryview/memoryview_annotation_typing.py", line 107, in memoryview_annotation_typing.one_dim_nogil_cfunc
Failed example:
    one_dim_nogil_cfunc(a)
Expected:
    2.0
Got:
    np.float64(2.0)


======================================================================
FAIL: two_dim (memoryview_annotation_typing)
Doctest: memoryview_annotation_typing.two_dim
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/tests/memoryview/memoryview_annotation_typing.py", line 39, in memoryview_annotation_typing.two_dim
Failed example:
    two_dim(a)
Expected:
    (3.0, 1.0, 2)
Got:
    (np.float64(3.0), np.float64(1.0), 2)


======================================================================
FAIL: variable_annotation (memoryview_annotation_typing)
Doctest: memoryview_annotation_typing.variable_annotation
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/tests/memoryview/memoryview_annotation_typing.py", line 49, in memoryview_annotation_typing.variable_annotation
Failed example:
    variable_annotation(a)
Expected:
    2.0
Got:
    np.float64(2.0)


======================================================================
ERROR: runTest (__main__.CythonRunTestCase.runTest)
[1] compiling (c/cy2) and running numpy_memoryview
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cython/cython/runtests.py", line 1493, in run
    ext_so_path = self.runCompileTest()
                  ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1108, in runCompileTest
    return self.compile(
           ^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1412, in compile
    self._match_output(expected_errors, errors, tostderr)
  File "/home/runner/work/cython/cython/runtests.py", line 1466, in _match_output
    self.assertEqual(None, unexpected)
AssertionError: None != "678:9: 'int_t' is not a type identifier"


======================================================================
FAIL: test_flow_control (ufunc)
Doctest: ufunc.test_flow_control
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_flow_control
Failed example:
    return_in_if(-1.)
Expected:
    1.0
Got:
    np.float64(1.0)
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_flow_control
Failed example:
    return_in_if(2.0)
Expected:
    2.0
Got:
    np.float64(2.0)
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_flow_control
Failed example:
    nested_loops(5.5)
Expected:
    6.0
Got:
    np.float64(6.0)
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_flow_control
Failed example:
    nested_loops(105.)
Expected:
    -5.0
Got:
    np.float64(-5.0)


======================================================================
FAIL: test_nested_function (ufunc)
Doctest: ufunc.test_nested_function
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_nested_function
Failed example:
    nested_function(-1.)
Expected:
    -2.0
Got:
    np.float64(-2.0)


======================================================================
FAIL: test_plus_one (ufunc)
Doctest: ufunc.test_plus_one
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/c/ufunc/ufunc.cpython-312-x86_64-linux-gnu.so", line ?, in ufunc.test_plus_one
Failed example:
    plus_one(1.j)
Expected:
    (1+1j)
Got:
    np.complex128(1+1j)


======================================================================
ERROR: runTest (__main__.CythonRunTestCase.runTest)
[0] compiling (c/cy2) and running numpy_test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cython/cython/runtests.py", line 1493, in run
    ext_so_path = self.runCompileTest()
                  ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1108, in runCompileTest
    return self.compile(
           ^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1412, in compile
    self._match_output(expected_errors, errors, tostderr)
  File "/home/runner/work/cython/cython/runtests.py", line 1466, in _match_output
    self.assertEqual(None, unexpected)
AssertionError: None != '257:29: Invalid type.'


======================================================================
FAIL: test_one_sized (relaxed_strides)
Doctest: relaxed_strides.test_one_sized
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/5/memoryview/c/relaxed_strides/relaxed_strides.cpython-312-x86_64-linux-gnu.so", line ?, in relaxed_strides.test_one_sized
Failed example:
    test_one_sized(contig)[0]
Expected:
    1.0
Got:
    np.float64(1.0)


======================================================================
FAIL: copy_fortran2 (contig_check)
Doctest: contig_check.copy_fortran2
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran2
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran2
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran2
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_


======================================================================
FAIL: copy_fortran3 (contig_check)
Doctest: contig_check.copy_fortran3
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran3
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran3
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/contig_check/contig_check.cpython-312-x86_64-linux-gnu.so", line ?, in contig_check.copy_fortran3
Failed example:
    (a == c).all()
Expected:
    True
Got:
    np.True_


======================================================================
FAIL: test_update_mmview_rw (numpy_memoryview_readonly)
Doctest: numpy_memoryview_readonly.test_update_mmview_rw
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/4/memoryview/c/numpy_memoryview_readonly/numpy_memoryview_readonly.cpython-312-x86_64-linux-gnu.so", line ?, in numpy_memoryview_readonly.test_update_mmview_rw
Failed example:
    test_update_mmview_rw(new_array())
Expected:
    23.0
Got:
    np.float64(23.0)


======================================================================
ERROR: runTest (__main__.CythonRunTestCase.runTest)
[6] compiling (c/cy2) and running numpy_parallel
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/cython/cython/runtests.py", line 1493, in run
    ext_so_path = self.runCompileTest()
                  ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1108, in runCompileTest
    return self.compile(
           ^^^^^^^^^^^^^
  File "/home/runner/work/cython/cython/runtests.py", line 1412, in compile
    self._match_output(expected_errors, errors, tostderr)
  File "/home/runner/work/cython/cython/runtests.py", line 1466, in _match_output
    self.assertEqual(None, unexpected)
AssertionError: None != '25:23: Invalid type.'
```
For the `repr()` differences, we can probably cast to the normal Python types using `float()` or `bool()`.

### Code to reproduce the behaviour:

_No response_

### Expected behaviour

_No response_

### OS

_No response_

### Python version

3.12

### Cython version

master

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattip**
Fixes #6072 

Should this also add a numpy2.0b1 test run to CI via the wheels at https://pypi.anaconda.org/scientific-python-nightly-wheels/numpy (i.e. install numpy via this index)?
```
pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy
```

Edit: make it a question

### CODE REVIEW

**Comment by mattip** [comment]
This is better but it does not fix everything. There is some mess since on NumPy1
```
typedef struct { double real, imag; } npy_cdouble;
typedef struct { double real, imag; } npy_cdouble;
typedef struct { npy_longdouble real, imag; } npy_clongdouble;
```

but on NumPy2
```
#if defined(__cplusplus)

typedef struct
{
    double _Val[2];
} npy_cdouble;

typedef struct
{
    float _Val[2];
} npy_cfloat;

typedef struct
{
    long double _Val[2];
} npy_clongdouble;

#else
#include <complex.h>
#if defined(_MSC_VER) && !defined(__INTEL_COMPILER)
typedef _Dcomplex npy_cdouble;
typedef _Fcomplex npy_cfloat;
typedef _Lcomplex npy_clongdouble;
#else /* !defined(_MSC_VER) || defined(__INTEL_COMPILER) */
typedef double _Complex npy_cdouble;
typedef float _Complex npy_cfloat;
typedef longdouble_t _Complex npy_clongdouble;
#endif
#endif
```

This is making compilation fail since cython assumes `.real` and `.imag` can be used on complex types.

**Comment by mattip** [comment]
There is no simple way to map `np.int_` to a C type on windows, since on NumPy1 it is an alias of `np.int32` but on NumPy2 it is an alias to `np.int64`. On non-windows, it is always an alias to `np.int64`.

**Comment by da-woods** [comment]
> There is no simple way to map `np.int_` to a C type on windows, since on NumPy1 it is an alias of `np.int32` but on NumPy2 it is an alias to `np.int64`. On non-windows, it is always an alias to `np.int64`.

I think dropping `np.int_` is reasonable. Or possibly running the test selectively on Numpy2 only.

Not sure what to do about the complex stuff. There probably isn't a sensible way of writing one test for Cython that covers both (because the interface has changed, and what we're really testing in Cython is the interface)

**Comment by mattip** [comment]
> Not sure what to do about the complex stuff.

I think cython should wait for a resolution to  numpy/numpy#26029, and adjust the tests appropriately. Clearly `arr[1].imag += 1` is wrong, but I am not sure what is right. I would like to have a way to write `np_csetimag(np_cimag(arr[1]), np_cimag(arr[1]) + 1)` more compactly.

In any case, this PR goes most of the way to fixing the incompatibility in a way that works on both versions, so if CI agrees maybe it could be merged and the rest handled in a subsequent PR.

**Comment by mattip** [comment]
CI is passing.

**Comment by da-woods** [comment]
> In any case, this PR goes most of the way to fixing the incompatibility in a way that works on both versions, so if CI agrees maybe it could be merged and the rest handled in a subsequent PR.

Yeah I think I'm happy to go along with this.

I'll put off a decision about backporting it to now and just merge it to master I think.

Thanks @mattip

**Comment by da-woods** [comment]
3.0.x commit e4b1e5ea653962dcf56327e075c0cc1befe16504 (since it looks like the Python 3.12 tests there use Numpy 2)

**Comment by mattip** [comment]
> 3.0.x commit https://github.com/cython/cython/commit/e4b1e5ea653962dcf56327e075c0cc1befe16504

I think #6100 is also needed

**Comment by da-woods** [comment]
> > 3.0.x commit [e4b1e5e](https://github.com/cython/cython/commit/e4b1e5ea653962dcf56327e075c0cc1befe16504)
> 
> I think #6100 is also needed

Yes I think so too - I still need to have a proper look at that

File: tests/run/numpy_parallel.pyx [hunk 1]
```diff
@@ -22,7 +22,7 @@ def test_parallel_numpy_arrays():
     4
     """
     cdef Py_ssize_t i, length
-    cdef np.ndarray[np.int_t] x
+    cdef np.ndarray[np.int64_t] x
```
Comment (stratakis) [inline tests/run/numpy_parallel.pyx:25]: On numpy 1.26.0 on i686 this test fails with:

```
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.12/doctest.py", line 1361, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest numpy_parallel.test_parallel_numpy_arrays[0]>", line 1, in <module>
        test_parallel_numpy_arrays()
      File "tests/run/numpy_parallel.pyx", line 34, in numpy_parallel.test_parallel_numpy_arrays (numpy_parallel.c:4494)
        x = numpy.zeros(10, dtype=numpy.int_)
    ValueError: Buffer dtype mismatch, expected 'int64_t' but got 'long'
```

**Pull request merged**